client/components/web-preview/toolbar.jsx

```
@@ -68,17 +80,30 @@ class PreviewToolbar extends Component {
 		this.props.onEdit();
 	};
 
-	handleEditorWebPreviewEditHeader = ( event ) => {
+	handleEditorWebPreviewEditHeader = async ( event ) => {
 		event.preventDefault();
+		this.setState( { isRedirecting: true } );
+
 		this.props.recordTracksEvent( 'calypso_editor_preview_edit_header_click' );
-		window.location.href = this.props.customizeUrl;
+
+		const { isAtomic, selectedSiteId, siteEditorUrl, themeInstallId } = this.props;
+
+		// For atomic sites, we need to install theme before navigating to site editor
+		// If theme is already installed, installation will silently fail, and we just switch to the site-editor.
+		try {
+			if ( isAtomic ) {
+				await this.props.installTheme( themeInstallId, selectedSiteId );
+			}
+			window.location.href = siteEditorUrl;
+		} catch ( error ) {
+			this.setState( { isRedirecting: false } );
+		}
```

Nitpick (non-blocking): Maybe we can check whether the theme is installed and avoid an unnecessary install action if the theme is already installed?

---

@mmtr I thought about this initially. Looking at the code, `installTheme` is already handling the case of installed themes correctly. When a theme is already installed, the request simply returns:

```
    "body": {
        "error": "theme_already_installed",
        "message": "The theme is already installed"
    }
```

and we proceed with the redirect to site-editor as expected.

Adding an additional check like `!!getTheme(state, siteId, ownProps.themeId)` would require checking this on all sites, even non-atomic ones, or we'd end up with nested conditionals before proceeding with `installTheme`. Since it's already handled properly to silently fail and redirect to the site editor, the extra check seems unnecessary.

The catch block here is specifically for handling actual issues that could arise during theme installation, such as network issues, server errors, or permission problems.